File: src/main/java/com/nerdearla/workshop/payment/PaymentService.java
package com.nerdearla.workshop.payment;

import com.nerdearla.workshop.authorization.PaymentAuthorization;
import com.nerdearla.workshop.operation.PaymentOperation;
import com.nerdearla.workshop.payment.model.Payment;
import com.nerdearla.workshop.payment.provider.PaymentIdProvider;
import com.nerdearla.workshop.payment.repository.PaymentRepository;
import com.nerdearla.workshop.paymentMethod.BuyerPaymentMethod;
import com.nerdearla.workshop.paymentMethod.PaymentMethodService;
import com.nerdearla.workshop.authorization.AuthorizationService;
import com.nerdearla.workshop.user.buyer.Buyer;
import com.nerdearla.workshop.user.buyer.BuyerService;
import com.nerdearla.workshop.qr.QR;
import com.nerdearla.workshop.qr.QRService;
import com.nerdearla.workshop.user.seller.Seller;
import com.nerdearla.workshop.user.seller.SellerService;
import org.springframework.stereotype.Service;

@Service
public class PaymentService {

    private final PaymentIdProvider paymentIdProvider;
    private final PaymentMethodService paymentMethodService;
    private final AuthorizationService gatewayService;
    private final BuyerService buyerService;
    private final SellerService sellerService;
    private final QRService qrService;
    private final PaymentRepository paymentRepository;

    public PaymentService(
            PaymentIdProvider paymentIdProvider,
            PaymentMethodService paymentMethodService,
            AuthorizationService gatewayService,
            BuyerService buyerService,
            SellerService sellerService,
            PaymentRepository paymentRepository,
            QRService qrService
    ) {
        this.paymentIdProvider = paymentIdProvider;
        this.paymentMethodService = paymentMethodService;
        this.gatewayService = gatewayService;
        this.buyerService = buyerService;
        this.sellerService = sellerService;
        this.paymentRepository = paymentRepository;
        this.qrService = qrService;
    }

    public Payment processPayment(PaymentOperation operation) {
        String id = paymentIdProvider.getNext();
        operation.setPaymentId(id);

        // validar qr_id?
        QR qr = qrService.findValidQR(operation.getPaymentRequest().getQrId());
        operation.setQr(qr);

        Buyer buyer = buyerService.findBuyer(operation.getPaymentRequest().getBuyerId());
        operation.setBuyer(buyer);

        // Obtener paymentMethod y agregarlo a operation?
        BuyerPaymentMethod buyerPaymentMethod = paymentMethodService.getBy(buyer.getId(), operation.getPaymentRequest().getPaymentMethodData());
        operation.setPaymentMethod(buyerPaymentMethod);

        Seller seller = sellerService.findSeller(operation.getPaymentRequest().getSellerId());
        operation.setSeller(seller);

        PaymentAuthorization authorization = gatewayService.authorize(operation);

        Payment payment = new Payment.PaymentBuilder(operation, authorization).build();
        paymentRepository.save(payment);

        return payment;
    }
}

File: src/main/java/com/nerdearla/workshop/payment/PaymentService.kt
package com.nerdearla.workshop.payment

import com.nerdearla.workshop.authorization.GatewayService
import com.nerdearla.workshop.operation.AuthorizedOperation
import com.nerdearla.workshop.operation.ExpandedOperation
import com.nerdearla.workshop.operation.ExpandedOperationProvider
import com.nerdearla.workshop.operation.InitialOperation
import com.nerdearla.workshop.payment.model.Payment
import com.nerdearla.workshop.payment.repository.PaymentRepository
import com.nerdearla.workshop.shared.utils.CompanionLogger
import org.springframework.stereotype.Service

@Service
class PaymentService(
    private val expandedOperationProvider: ExpandedOperationProvider,
    private val gatewayService: GatewayService,
    private val paymentRepository: PaymentRepository
) {
    //quitamos el nulo de la respuesta

    // copy vs construccion con todos los datos
    // extensions vs let / also, pros y contras
    fun processPayment(initialOperation: InitialOperation) =
        initialOperation
            .expandEntities()
            .authorize()
            .toPayment()
            .also { save(it) }

    private fun InitialOperation.expandEntities() =
        expandedOperationProvider.provide(this)
            .log { info("operation expanded {}", it) }

    private fun ExpandedOperation.authorize() =
        gatewayService.authorize(this)
            .log { info("authorization response: {}", it) }
            .let { AuthorizedOperation(this, it.id) }

    private fun AuthorizedOperation.toPayment() =
        Payment(
            id = operation.paymentId,
            amount = operation.amount,
            authorizationId = authorizationId,
            buyerId = operation.buyer.id,
            sellerId = operation.seller.id,
            paymentMethodId = operation.buyerPaymentMethod.id,
            qrId = operation.qr.id,
        )

    private fun save(payment: Payment) =
        paymentRepository.save(payment)
            .log { info("payment {} saved", payment.id) }

    companion object : CompanionLogger()
}
